backend/space/models.py

```diff
@@ -0,0 +1,194 @@
+        to="Structure",
+        through="StructureField",
+        help_text="sites referenced in this description",
+    )
+
+    def __str__(self):
+        return self.name
+
+
+class NamedSpace(models.Model):
+    """
+    Abstract class for "Named" regions, i.e. ones that can be
+    identified as named entities.
+    """
+
+    name = models.CharField(
+        max_length=200,
+        unique=True,
+        blank=False,
+    )
+
+    description = models.TextField(
+        blank=True,
+    )
+
+    identifiable = models.BooleanField(
+        default=True,
+        null=False,
+        help_text="Whether this place is an identifiable location that can be cross-referenced between descriptions, or a generic description",
+    )
+
+    # may be expanded with geo data?
+
+    def __str__(self):
+        return self.name
+
+    class Meta:
+        abstract = True
+
+
+class PoliticalRegion(NamedSpace, models.Model):
+    """
+    A political region, e.g. a kingdom or duchy
+    """
+
+    pass
+
+
+class EcclesiasticalRegion(NamedSpace, models.Model):
+    """
+    An ecclesiastical region, e.g. a dioceses
+    """
+
+    pass
+
+
+class GeographicalRegion(NamedSpace, models.Model):
+    """
+    A geographical region or location.
+
+    Unlike political or ecclesiastical regions, geographic locations
+    may be recognised today, e.g. "the Pyrenees".
+    """
+
+    pass
+
+
+class Structure(NamedSpace, models.Model):
+    """
+    A structure is a man-made site.
+
+    This can be a population centre ("Poitiers"), fortification
+    ("the monastery of Toulouse"), building ("a church"), room,
+    road, or even a specific object ("a desk").
+    """
+
+    class LevelOptions(models.IntegerChoices):
+        SETTLEMENT = 0, "settlement, population centre"
+        ROAD = 1, "road, square, crossroad"
+        FORTIFICATION = 2, "fortification"
+        BUILDING = 3, "building, vessel"
+        ROOM = 4, "room"
+        SPOT = 5, "spot, object"
+
+    level = models.IntegerField(choices=LevelOptions.choices)
+    parent = models.ForeignKey(
+        to="self",
+        related_name="children",
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        verbose_name="parent structure",
+        help_text="The structure containing this structure, e.g. the building containing a room.",
+    )
+
+    @admin.display()
+    def ancestors(self):
+        if self.parent:
+            return [self.parent] + self.parent.ancestors()
+        else:
+            return []
+
+    @admin.display()
+    def descendants(self):
+        iterate_descendants = (
+            [child] + child.descendants() for child in self.children.all()
+        )
+        return list(itertools.chain.from_iterable(iterate_descendants))
+
+    def clean(self):
+        if self.parent:
+            validators.validate_level_deeper_than_parent(
+                self.level, self.parent, self.LevelOptions
+            )
+
```

Comment:
Maybe you could extract the names of ancestors/descendants so they show up nicely in the Admin instead of: `Descendants: [<Structure: St. Radegund's desk>, <Structure: St. Radegund's room>]`

Reply:
Done! They'll now display as `St. Radegund's desk, St. Radegund's room`.